src/librustc_trans/intrinsic.rs

```diff
@@ -138,6 +141,30 @@ pub fn trans_intrinsic_call<'a, 'tcx>(bx: &Builder<'a, 'tcx>,
             let llfn = cx.get_intrinsic(&("llvm.debugtrap"));
             bx.call(llfn, &[], None)
         }
+        "va_arg" => {
+            match ret_ty.sty {
+                // `va_arg` should never be used with the return type f32. If the
+                // return type is f32, use `va_arg` with the return type of f64
+                // followed by fptrunc to truncate the result to an f32.
+                ty::TyFloat(ast::FloatTy::F32) => {
+                    let ext_ret_ty = bx.va_arg(args[0].immediate(), Type::f64(cx));
+                    bx.fptrunc(ext_ret_ty, Type::f32(cx))
+                }
+                ty::TyBool |
+                ty::TyInt(_) |
+                ty::TyUint(_) |
+                ty::TyFloat(_) |
+                ty::TyRawPtr(_) => {
+                    bx.va_arg(args[0].immediate(), llret_ty)
```

Comment:
I wasn't sure which lint was referred to to make `va_arg` only work with ffi safe types. I listed them explicitly here. I also haven't figured out how to return a non-scalar type yet. LLVM keeps hitting a "Unknown type" assertion, so `TyAdt`s aren't listed here yet.

Reply:
You can't use `span_invalid_monomorphization_error` here - that's reserved for polymorphic intrinsics exposed via monomorphic APIs (such as a lot of the SIMD stuff).

The lint is around here (you'd want to check calls of `VaList::arg` instead of any definitions): https://github.com/rust-lang/rust/blob/ed9a29a976f8b73f3fc91a10d0b5addd86a394c7/src/librustc_lint/types.rs#L749

Also, I was led to believe non-primitives in varargs were supported by LLVM, but it sounds like that's not the case? If so, shouldn't a trait be used to *whitelist* the types allowed?

Reply:
Maybe if arbitrary types can be supported, you could check what `clang` does, in various cases? Or even the implementation in `clang`.

Reply:
I could be wrong but AIUI the `improper_ctypes` lint only checked extern blocks for functions we import, not functions we export. E.g.

```
// lint is not triggered
pub extern "C" fn exported_non_ffi_safe(x: SomeUnion) -> SomeZst { }

extern {
    fn imported_non_ffi_safe(x: SomeUnion) -> SomeZst; // lint is triggered
}
```

> You can't use span_invalid_monomorphization_error here - that's reserved for polymorphic intrinsics exposed via monomorphic APIs (such as a lot of the SIMD stuff).

Is a regular `tcx.sess.span_err` allowed?

> Also, I was led to believe non-primitives in varargs were supported by LLVM, but it sounds like that's not the case? If so, shouldn't a trait be used to whitelist the types allowed?

Same, Currently running `llc` on the following causes `llc` to segfault.
```
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

%struct.va_list_tag = type { i32, i32, {}*, {}* }
%struct.point = type { i64, i64 }

define %struct.point @va_list_test(i64, %struct.va_list_tag*) {
    %3 = bitcast %struct.va_list_tag* %1 to %struct.va_list_tag*
    %4 = va_arg %struct.va_list_tag* %3, %struct.point
    ret %struct.point %4
}
```

So I'm guessing we won't be able to just use `va_arg` for aggregate types. `clang` appears to use GEP.

Reply:
Definitely not simple to replicate well with sufficient testing.

Reply:
Given that extra complexity and how niche non-primitive types in variadic arguments are (even compared to variadic functions in general) and the fact that primitives includes "all thin pointers", I'd suggest punting on arbitrary user-defined types for now. With a sealed trait, you can ensure only known-good types can be used as variadic arguments, and if there's sufficient motivation and implementation effort for lifting that restriction later, it can be lifted backwards compatibly.

This could be either an RFC amendment or just a known implementation limitation (of which we have plenty).

Reply:
> Given that extra complexity and how niche non-primitive types in variadic arguments are (even compared to variadic functions in general) and the fact that primitives includes "all thin pointers", I'd suggest punting on arbitrary user-defined types for now.

Agreed. I'll start working on this.